File: packages/ui/src/standards/symbol-generators/fire-safety.ts
import type { FireSafetyElement } from '../../services/kernel-bridge'
import type { SymbolGenerationOptions, SymbolPrimitiveBundle } from '../symbol-primitives'
import { getDefaultSymbolProvenance, getSymbolSpec } from '../plan-symbol-profile'
import {
  addRect, addLine, addCircle, addText,
  type SymbolPrimitive,
} from './_helpers'

function textForType(symbolType: FireSafetyElement['symbol_type']): string | null {
  switch (symbolType) {
    case 'fire_extinguisher':
      return 'FE'
    case 'pull_station':
      return 'FP'
    case 'smoke_alarm':
      return 'SA'
    case 'fire_alarm_panel':
      return 'FACP'
    case 'fire_hose_cabinet':
      return 'FHC'
    case 'annunciator':
      return 'ANN'
    default:
      return null
  }
}

export function generateFireSafetyPrimitives(
  element: FireSafetyElement,
  options: SymbolGenerationOptions,
): SymbolPrimitiveBundle {
  const spec = getSymbolSpec(options.planSymbolProfile, 'fire_safety', element.symbol_type) ?? {
    domain: 'fire_safety' as const,
    title: element.symbol_type,
    lineClass: 'primary' as const,
    allowsLabel: false,
    provenance: getDefaultSymbolProvenance(),
  }
  const primitives: SymbolPrimitive[] = []
  const [x, y] = element.position
  const rotation = element.rotation ?? 0

  switch (element.symbol_type) {
    case 'fire_extinguisher': {
      addCircle(primitives, [x, y], rotation, [0, 0], 0.10, 'primary')
      addText(primitives, [x, y], rotation, 'FE', 0.07)
      break
    }

    case 'sprinkler_head': {
      // Outer circle
      addCircle(primitives, [x, y], rotation, [0, 0], 0.08, 'primary')
      // Center dot
      addCircle(primitives, [x, y], rotation, [0, 0], 0.02, 'primary')
      // 4 short radiating lines at 0, PI/2, PI, 3PI/2
      const innerR = 0.04
      const outerR = 0.09
      addLine(primitives, [x, y], rotation, [innerR, 0], [outerR, 0], 'secondary')
      addLine(primitives, [x, y], rotation, [0, innerR], [0, outerR], 'secondary')
      addLine(primitives, [x, y], rotation, [-innerR, 0], [-outerR, 0], 'secondary')
      addLine(primitives, [x, y], rotation, [0, -innerR], [0, -outerR], 'secondary')
      break
    }

    case 'exit_sign': {
      addRect(primitives, [x, y], rotation, 0.30, 0.12, 'primary')
      addText(primitives, [x, y], rotation, 'EXIT', 0.06)
      // Direction arrow to the right
      addLine(primitives, [x, y], rotation, [0.18, 0], [0.24, 0], 'secondary')
      addLine(primitives, [x, y], rotation, [0.21, 0.03], [0.24, 0], 'secondary')
      addLine(primitives, [x, y], rotation, [0.21, -0.03], [0.24, 0], 'secondary')
      break
    }

    case 'pull_station': {
      addRect(primitives, [x, y], rotation, 0.12, 0.12, 'primary')
      addText(primitives, [x, y], rotation, 'FP', 0.06)
      break
    }

    case 'smoke_alarm': {
      addCircle(primitives, [x, y], rotation, [0, 0], 0.09, 'primary')
      addText(primitives, [x, y], rotation, 'SA', 0.07)
      break
    }

    case 'fire_alarm_panel': {
      addRect(primitives, [x, y], rotation, 0.30, 0.20, 'primary')
      addText(primitives, [x, y], rotation, 'FACP', 0.05)
      break
    }

    case 'fire_hose_cabinet': {
      addRect(primitives, [x, y], rotation, 0.60, 0.20, 'primary')
      addText(primitives, [x, y], rotation, 'FHC', 0.06)
      break
    }

    case 'annunciator': {
      addRect(primitives, [x, y], rotation, 0.40, 0.10, 'primary')
      addText(primitives, [x, y], rotation, 'ANN', 0.05)
      break
    }

    default: {
      addCircle(primitives, [x, y], rotation, [0, 0], 0.10, spec.lineClass as 'primary')
      break
    }
  }

  const tag = textForType(element.symbol_type)
  const showTag = Boolean(tag) && (options.showLabels || options.annotationDensity !== 'minimal')
  if (showTag && tag) {
    addText(primitives, [x, y], rotation, tag, 0.12)
  }

  return {
    domain: spec.domain,
    primitives,
  }
}

File: crates/bcad-2d/src/symbols/fire_safety.rs
//! Fire safety symbol generators.
//!
//! Geometry closely mirrors the TypeScript generators in
//! `packages/ui/src/standards/symbol-generators/fire-safety.ts`.

use crate::style::{DomainColor, StyledPrimitive, SymbolLineClass};

use super::{add_circle, add_line, add_rect, add_text};

const DOMAIN: DomainColor = DomainColor::FireSafety;

/// Fire extinguisher: circle + "FE" label.
pub fn fire_extinguisher(center: [f64; 2], rotation: f64) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    add_circle(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [0.0, 0.0],
        0.10,
        SymbolLineClass::Primary,
        false,
    );
    add_text(&mut out, DOMAIN, center, rotation, "FE", 0.07);
    out
}

/// Sprinkler head: outer circle + center dot + 4 radiating lines.
pub fn sprinkler_head(center: [f64; 2], rotation: f64) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    add_circle(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [0.0, 0.0],
        0.08,
        SymbolLineClass::Primary,
        false,
    );
    add_circle(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [0.0, 0.0],
        0.02,
        SymbolLineClass::Primary,
        false,
    );
    let inner_r = 0.04;
    let outer_r = 0.09;
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [inner_r, 0.0],
        [outer_r, 0.0],
        SymbolLineClass::Secondary,
        false,
    );
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [0.0, inner_r],
        [0.0, outer_r],
        SymbolLineClass::Secondary,
        false,
    );
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [-inner_r, 0.0],
        [-outer_r, 0.0],
        SymbolLineClass::Secondary,
        false,
    );
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [0.0, -inner_r],
        [0.0, -outer_r],
        SymbolLineClass::Secondary,
        false,
    );
    out
}

/// Exit sign: rectangle + "EXIT" label + direction arrow.
pub fn exit_sign(center: [f64; 2], rotation: f64) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    add_rect(
        &mut out,
        DOMAIN,
        center,
        rotation,
        0.30,
        0.12,
        SymbolLineClass::Primary,
        [0.0, 0.0],
        false,
    );
    add_text(&mut out, DOMAIN, center, rotation, "EXIT", 0.06);
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [0.18, 0.0],
        [0.24, 0.0],
        SymbolLineClass::Annotation,
        false,
    );
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [0.21, 0.03],
        [0.24, 0.0],
        SymbolLineClass::Annotation,
        false,
    );
    add_line(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [0.21, -0.03],
        [0.24, 0.0],
        SymbolLineClass::Annotation,
        false,
    );
    out
}

/// Pull station: small square + "FP" label.
pub fn pull_station(center: [f64; 2], rotation: f64) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    add_rect(
        &mut out,
        DOMAIN,
        center,
        rotation,
        0.12,
        0.12,
        SymbolLineClass::Primary,
        [0.0, 0.0],
        false,
    );
    add_text(&mut out, DOMAIN, center, rotation, "FP", 0.06);
    out
}

/// Smoke alarm: circle + "SA" label.
pub fn smoke_alarm(center: [f64; 2], rotation: f64) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    add_circle(
        &mut out,
        DOMAIN,
        center,
        rotation,
        [0.0, 0.0],
        0.09,
        SymbolLineClass::Primary,
        false,
    );
    add_text(&mut out, DOMAIN, center, rotation, "SA", 0.07);
    out
}

/// Fire alarm panel: rectangle + "FACP" label.
pub fn fire_alarm_panel(center: [f64; 2], rotation: f64) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    add_rect(
        &mut out,
        DOMAIN,
        center,
        rotation,
        0.30,
        0.20,
        SymbolLineClass::Primary,
        [0.0, 0.0],
        false,
    );
    add_text(&mut out, DOMAIN, center, rotation, "FACP", 0.05);
    out
}

/// Fire hose cabinet: wide rectangle + "FHC" label.
pub fn fire_hose_cabinet(center: [f64; 2], rotation: f64) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    add_rect(
        &mut out,
        DOMAIN,
        center,
        rotation,
        0.60,
        0.20,
        SymbolLineClass::Primary,
        [0.0, 0.0],
        false,
    );
    add_text(&mut out, DOMAIN, center, rotation, "FHC", 0.06);
    out
}

/// Annunciator: wide thin rectangle + "ANN" label.
pub fn annunciator(center: [f64; 2], rotation: f64) -> Vec<StyledPrimitive> {
    let mut out = Vec::new();
    add_rect(
        &mut out,
        DOMAIN,
        center,
        rotation,
        0.40,
        0.10,
        SymbolLineClass::Primary,
        [0.0, 0.0],
        false,
    );
    add_text(&mut out, DOMAIN, center, rotation, "ANN", 0.05);
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn sprinkler_head_produces_primitives() {
        let prims = sprinkler_head([0.0, 0.0], 0.0);
        // 2 circles + 4 lines = 6
        assert!(prims.len() >= 6);
    }

    #[test]
    fn exit_sign_produces_primitives() {
        let prims = exit_sign([0.0, 0.0], 0.0);
        // rect + text + 3 arrow lines = 5
        assert!(prims.len() >= 5);
    }
}
